sinks/kafka.go

```diff
@@ -283,68 +292,92 @@ func startProducer(conf *kafkaSinkConfig, in chan *sarama.ProducerMessage, stop
 		lastSend, timepass int64
+		msgsSize, msgSize  int
 	)
 	lastSend = time.Now().Unix()
-ProducerLoop:
 	for {
 		select {
 		case msg = <-in:
 			if msg.Value.Length() != 0 {
-				msgs = append(msgs, msg)
-				enqueued++
-				timepass = time.Now().Unix() - lastSend
-				if timepass >= 1 || len(msgs) >= conf.BatchSize {
-					err := producer.SendMessages(msgs)
-					if err != nil {
-						producerErr := err.(sarama.ProducerErrors)
-						for len(producerErr) > 0 && retries < 10 {
-							log.Warn("basicProducer: failed to push to kafka, try to resend")
-							//TODO check if fisrt error is first event
-							for i, v := range msgs {
-								if v.Value == producerErr[0].Msg.Value {
-									msgs = msgs[i:]
-								}
-							}
-							retries++
-							//resend
-							err = producer.SendMessages(msgs)
-							if err != nil {
-								producerErr = err.(sarama.ProducerErrors)
-							} else {
-								continue
-							}
-
-						}
-						if retries >= 20 {
-							log.WithFields(log.Fields{
-								"nbRetry": retries,
-							}).Panic("Failed to push event to kafka. Stopping agent.")
-						}
-					}
-
-					log.WithFields(log.Fields{
-						"size":     len(msgs),
-						"Timepass": timepass,
-						"enqueued": enqueued,
-					}).Debug("basic producer length")
-					msgs = []*sarama.ProducerMessage{}
-					retries = 0
-					enqueued = 0
 
+				//calcul size
+				msgSize = msgByteSize(msg)
+				if msgSize > k.kafkaConf.MaxMessageBytes {
+					log.Warn("Skip Message")
+
+				} else if msgsSize+msgSize < k.kafkaConf.MaxMessageBytes {
+					msgs = append(msgs, msg)
+					msgsSize += msgSize
+				} else {
+					lastSend = sendMsg(msgs, producer)
+					msgs = []*sarama.ProducerMessage{}
+					msgs = append(msgs, msg)
+					msgsSize = msgSize
 				}
+
 			}
-		case <-stop:
+
+			//use to clear slice
+			now := time.Now().Unix()
+			timepass = now - lastSend
+			if timepass >= 1 {
+				lastSend = sendMsg(msgs, producer)
+				msgs = []*sarama.ProducerMessage{}
+				msgsSize = 0
+			}
+
+		case <-k.stop:
 			log.Info("startProducer: Signal received, closing producer")
-			break ProducerLoop
+			return
 		}
 	}
-	log.WithFields(log.Fields{
-		"Enqueued": enqueued,
-	}).Info("startProducer")
+}
+
+func sendMsg(msgs []*sarama.ProducerMessage, producer sarama.SyncProducer) int64 {
+	retries := 0
+	err := producer.SendMessages(msgs)
+	for err != nil {
+		producerErrs := err.(sarama.ProducerErrors)
+		msgs = []*sarama.ProducerMessage{}
+		for _, v := range producerErrs {
+			log.WithFields(log.Fields{
+				"error": v.Err,
+			}).Warn("failed to push to kafka")
+			msgs = append(msgs, v.Msg)
+		}
+
+		if retries > 20 {
+			log.WithFields(log.Fields{
+				"nbRetry": retries,
+			}).Panic("Failed to push event to kafka. Stopping agent.")
+		}
+		retries++
+		err = producer.SendMessages(msgs)
+	}
+	return time.Now().Unix()
+}
+
+func msgByteSize(msg *sarama.ProducerMessage) int {
+	// the metadata overhead of CRC, flags, etc.
+	size := 26
```

Comment:
26 : Magic Number ? Add a comment to explain why it's 26. If if has maning, put it into a const or set as configurable

Reply:
see comment : the metadata overhead of CRC, flags, etc.

Reply:
I prefer keep like that